pkg/fileutils/fileutils.go

```diff
@@ -1,33 +1,77 @@
 package fileutils
 
 import (
+	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/Sirupsen/logrus"
 )
 
 // Matches returns true if relFilePath matches any of the patterns
+// and isn't excluded by any of the subsequent patterns.
+func exclusion(pattern string) bool {
+	return pattern[0] == '!'
+}
+
+func empty(pattern string) bool {
+	return pattern == ""
+}
+
 func Matches(relFilePath string, patterns []string) (bool, error) {
-	for _, exclude := range patterns {
-		matched, err := filepath.Match(exclude, relFilePath)
+	matched := false
+
+	for _, pattern := range patterns {
+		pattern = filepath.Clean(pattern)
+		negative := false
+		if empty(pattern) {
+			continue
+		}
+
+		if exclusion(pattern) {
+			if len(pattern) == 1 {
+				logrus.Errorf("Illegal exclusion pattern: %s", pattern)
+				return false, errors.New("Illegal exclusion pattern: !")
+			}
+			negative = true
+			pattern = pattern[1:]
+		}
+
+		match, err := filepath.Match(pattern, relFilePath)
 		if err != nil {
-			logrus.Errorf("Error matching: %s (pattern: %s)", relFilePath, exclude)
+			logrus.Errorf("Error matching: %s (pattern: %s)", relFilePath, pattern)
 			return false, err
 		}
-		if matched {
+		if match {
 			if filepath.Clean(relFilePath) == "." {
-				logrus.Errorf("Can't exclude whole path, excluding pattern: %s", exclude)
-				continue
+				logrus.Errorf("Can't exclude whole path")
+				return false, nil
+			}
+		} else if !match {
+			// Check to see if the pattern matches one of our parent dirs
+			path := filepath.Dir(relFilePath)
```

Comment:
these two lines can probably move out of the loop so that they're only done once. Sorry, I should have done more clean-up last night but I was low on time and just wanted to see if the general idea was even possible.